[pythonscripting] current marketplace version of pythonscripting next (#19190)
openhab/openhab-addons: bundles/org.openhab.automation.pythonscripting/README.md

        self.logger.info("Rule was triggered")
```

### `PY` Transformation

Or as transformation inline script

```text
String Test "Test [PY(|'String has ' + str(len(input)) + 'characters'):%s]"
```

### More Scripting

A complete Documentation about Python Scripting Rules and Transformation Scripts can be found at

> [>> openHAB Python Scripting <<](https://github.com/openhab/openhab-python/blob/main/README.md)

Reviewer: Would be nice to integrate this into the openHAB website where the other docs reside. We can take care of that later though, please remind me though.

Author: The idea was to split the binding documentation from the helper lib documentation.

Both documentations together was getting longer and longer. I still gave some scripting examples in the binding documentation itself, to get an impression how it works. But the whole helper lib documentation should resist in the same repo as the helper lib itself. Before, I was always synchronizing the README's after each tiny change in the helper libs, which affects the README.  Sometimes, I forgot to synchronize.

including all examples above, much more detailed.

## Add-on Administration

### Configuration

Web based config dialog can be found via Web UI => Settings / Add-on Settings / Python Scripting

![Pythonscripting configuration](doc/pythonscripting_configuration.png)

Additionally, you can configure the Add-on via a config file `/openhab/services/pythonscripting.cfg` like below.

> ```

Reviewer: ```suggestion
```text
```

Author: done

# Enable dependency tracking
#
# Dependency tracking allows your scripts to automatically reload when one of its dependencies is updated.
# You may want to disable dependency tracking if you plan on editing or updating a shared library, but don't want all
# your scripts to reload until you can test it.
#
#org.openhab.automation.pythonscripting:dependencyTrackingEnabled = true

# Cache compiled openHAB Python modules (.pyc files)
#
# Cache the openHAB python modules for improved startup performance.<br>
# Disable this option will result in a slower startup performance, because scripts have to be recompiled on every startup.
#
#org.openhab.automation.pythonscripting:cachingEnabled = true

# Enable jython emulation
#
# This enables Jython emulation in GraalPy. It is strongly recommended to update code to GraalPy and Python 3 as the emulation can have performance degradation.
# For tips and instructions, please refer to <a href="https://www.graalvm.org/latest/reference-manual/python/Modern-Python-on-JVM">Jython Migration Guide</a>.
#
#org.openhab.automation.pythonscripting:jythonEmulation = false
```

::: tip Configuration note

> If you use the marketplace version of this Add-on, it is neccessary to use the config file. OpenHAB has a bug which prevents the web based config dialog to work correctly for `kar` file based Add-ons.

Reviewer: Is there an issue in the webui repo for this?

> If you use the marketplace version of this Add-on, it is neccessary to use the config file. OpenHAB has a bug which prevents the web based config dialog to work correctly for `kar` file based Add-ons.

Author: done

:::

### Console

The [openHAB Console](https://www.openhab.org/docs/administration/console.html) provides access to additional features of these Add-on.

1. `pythonscripting info` is showing you additional data like version numbers, activated features and used path locations<br/><br/>![Pythonscripting info](doc/console_pythonscripting_info.png)

2. `pythonscripting console` provides an interactive python console where you can try live python features<br/><br/>![Pythonscripting console](doc/console_pythonscripting_console.png)

3. `pythonscripting update` allowes you to check, list, update or downgrade your helper lib<br/><br/>![Pythonscripting update](doc/console_pythonscripting_update.png)

> 4. `pythonscripting pip` allowes you check, install or remove external python modules.<br/><br/>Check [pip usage](#using-pip-to-install-external-modules) for details

Author: done

> 4. `pythonscripting pip` allowes you check, install or remove external python modules.<br/><br/>Check [pip usage](#using-pip-to-install-external-modules) for details

Author: done

    # zypper install patchelf
    # yum install patchelf
    ```

After these steps, venv setup is done and will be detected automatically during next openHAB restart.

::: tip VEnv note
Theoretically you can create venvs with a native python installation too. But it is strongly recommended to use graalpy for it. It will install a "special" version of pip in this venv, which will install patched python modules if available. This increases the compatibility of python modules with graalpython.

In container environments, you should mount the 'graalpy' folder to, because the venv is using symbolik links.
:::

### Using pip to install external modules

As first, you must enable [VEnv](#enabling-venv). After this is enabled, you can use pip in 2 ways.

1. Using the pythonscripting console<br/><br/>![Pythonscripting pip install](doc/console_pythonscripting_pip_install.png)![Pythonscripting pip install](doc/console_pythonscripting_pip_list.png)![Pythonscripting pip install](doc/console_pythonscripting_pip_show.png)

2. Using venv pip on your host system

    ```
    /openhab/userdata/cache/org.openhab.automation.pythonscripting/venv/bin/pip install requests
    ```

> ### Enable Pyhton Autocompletion

Reviewer: ```suggestion
### Enable Python Autocompletion
```

Author: done

Before you can enable autocompletion, you must generate the required type hint stub files. Login into [openHAB console](https://www.openhab.org/docs/administration/console.html) and run

```
pythonscripting typing
```

This will scan your current openHAB instance, including all installed Add-ons, for public java class methods and create corresponding python type hint stub files.

![Pythonscripting typing](doc/console_pythonscripting_typing.png)

The files are stored in the folder `/openhab/conf/automation/python/typings/`.

As a final step, the folders `/openhab/conf/automation/python/libs/` and `/openhab>/conf/automation/python/typings/` must be added as "extraPaths" in your IDE.

![Pythonscripting autocompletion](doc/ide_autocompletion.png)

## Typical log errors

### Graal python language not initialized. ...

```log
2025-07-25 12:10:06.001 [ERROR] [g.internal.PythonScriptEngineFactory] - Graal python language not initialized. Restart openhab to initialize available graal languages properly.
```

> This can happen after a new Add-on installation, if JSScripting is active at the same time.

Reviewer: ```suggestion
This can happen after a new Add-on installation, if JavaScript Scripting is active at the same time.
```

```

The reading the Python source file "conf/automation/python/lib/openhab/\_\_wrapper\_\_.py" failed.

This could either a permission/owner problem or a problem during deployment of the helper libs.
You should check that this file exists and it is readable by openHAB.
You should also check your logs for a message related to the helper lib deployment by just grep for "helper lib".

### SystemError, Option python.NativeModules is set to 'true' and a second GraalPy context ...

```log
SystemError, Option python.NativeModules is set to 'true' and a second GraalPy context attempted to load a native module '<xyz>' from path '<lib_path>.so'. At least one context in this process runs with 'IsolateNativeModules' set to false. Depending on the order of context creation, this means some contexts in the process cannot use native module, all other contexts must fall back and set python.NativeModules to 'false' to run native extensions in LLVM mode. This is recommended only for extensions included in the Python standard library. Running a 3rd party extension in LLVM mode requires a custom build of the extension and is generally discouraged due to compatibility reasons.
```

> These errors can occur if you use a native library in your external module but forgot to enable "native modules". Check the [Add-on configuration](#configuration) and enable 

Reviewer: ```suggestion
These errors can occur if you use a native library in your external module but forgot to enable "native modules". Check the [Add-on configuration](#configuration) and enable 
```
```suggestion
These errors can occur if you use a native library in your external module but forgot to enable "native modules". Check the [Add-on configuration](#configuration) and enable "native modules".
```

Author: done
